### ISSUE TITLE
Tracking Issue for `core::mem::copy`

### ISSUE THREAD

**Issue by jyn514** [rust-lang/rust#98262]
Feature gate: `#![feature(mem_copy_fn)]`

This is a tracking issue for `core::mem::copy`, which is a way - like `mem::drop` - to copy a value without creating a new closure.

### Public API

```rust
// core::mem
pub fn copy<T: Copy>(x: &T) -> T {}
```

### Steps / History

- [ ] Implementation: #95534
- [ ] Final comment period (FCP)
- [ ] Stabilization PR

### Unresolved Questions

- Should we find a way to make this an associated function on `Copy` instead?

**Comment by eggyal** [comment]
If you have `t: &T` where `T: Copy`, why would you want `core::mem::copy(t)` when you can instead do `*t`?

**Comment by jyn514** [comment]
@eggyal it's the difference between `option.map(|x| *x)` and `option.map(copy)`. The second is easier to read and write.

**Comment by DutchGhost** [comment]
Why not make it const right away?

```Rust
pub const fn copy<T: Copy>(x: &T) -> T {
    *x
}
```

compiles on stable...

**Comment by eggyal** [comment]
> Why not make it const right away?

Of course, this works if it's a free function; but would not (for now) if it's made an associated function on the `Copy` trait—so perhaps that alone might resolve the unresolved question ?

**Comment by Lokathor** [comment]
i'm mildly concerned that std::mem::copy and std::ptr::copy both would exist but do very different sorts of things. Can we call this new function `copied`, same as the iterator/option method?

**Comment by jyn514** [comment]
Hmm sure, copied seems fine.

I do think that all the functions in std::ptr are kind of expected to deal with raw memory; I don't think it would be hard to tell them apart in practice. But I don't feel strongly.

**Comment by Lokathor** [comment]
in an ideal world everything could be fully imported, so this helps the ideal.

**Comment by jyn514** [comment]
prior art: https://github.com/rust-lang/rust/blob/7b8e2a52ff29d400b4d869fad1d917ffa9bdce59/compiler/rustc_middle/src/ty/query.rs#L109-L113

**Comment by aleksanderkrauze** [comment]
I think this function should be marked `#[must_use]` as well.

**Comment by Stargateur** [comment]
I feel this method while useful may be premature, specially we already have other solution like `copied()` and `cloned()`, I advice we wait for a strong demand before make this a stable method. I don't think I ever needed such thing except for few case in Option or Iterator that already have solution for that. If this copy is not in the prelude (hiding in mem module), there is very few reason to have it, write the closure could be as fast. `Copy::copy` associated function could be better.

**Comment by KSBilodeau** [comment]
To throw my hat in the ring here, I actually ran into a helpful use case for this function today.  Using the unstable iter_next_chunk api on an vector of u8s, I got a `Result<[&u8; 4], ...>`, which cannot be converted easily to `[u8; 4]` with copied and either requires the length `.map(|x| *x)` or this lovely lil std::mem::copy function.

**Comment by zheland** [comment]
Regarding the use cases, the dereference operator does not always copy values implicitly. But sometimes it may be necessary to avoid re-borrowing and copy them instead to make them mutable.
In the example below `fn1`, `fn2`, `fn3`, `fn4` compile just fine, while `fn5` fails to compile. The use of `copy` (or the similar function/method) in `fn4` (*fixed) in this case seems quite explicit and clear.

```rust
fn fn1(x: &u8) { let mut x = *x; take_by_mut(&mut x); } // OK
fn fn2(x: &u8) { take_by_mut(&mut (|| *x)()); } // OK
fn fn3(x: &u8) { take_by_mut(&mut (*x,).0); } // OK
fn fn4(x: &u8) { take_by_mut(&mut copy(x)); } // OK

// ERROR: cannot borrow `*x` as mutable, as it is behind a `&` reference
// `x` is a `&` reference, so the data it refers to cannot be borrowed as mutable
fn fn5(x: &u8) { take_by_mut(&mut *x); }

fn take_by_mut(_: &mut u8) {}
const fn copy<T: Copy>(x: &T) -> T { *x }
```

**Comment by Skgland** [comment]
> The use of `copy` (or the similar function/method) in `fn3` in this case seems quite explicit and clear.

`fn4`?

Another possibility is `fn fn6(x: &u8) { take_by_mut(&mut {*x}); }` (the block forces place-to-value coercion ?), simpler than some of the other examples, though still less explicit and less clear than usning a `copy` function.

**Comment by suguruwataru** [comment]
I was looking for a non-transmute version of `core::mem::transmute_copy` and came across this, so the naming may create a bit of confusion... I feel an unsafe `copy` which does a single value `memcpy` fits in `core::mem` better than a shorthand for `|x| *x`.

**Comment by eggyal** [comment]
@suguruwataru isn't that just [`std::ptr::read`](https://doc.rust-lang.org/std/ptr/fn.read.html) (albeit with the added safety of taking a shared reference rather than a raw pointer, however the former coerces to the latter so it makes little ergonomic difference)?

**Comment by suguruwataru** [comment]
> [@suguruwataru](https://github.com/suguruwataru) isn't that just [`std::ptr::read`](https://doc.rust-lang.org/std/ptr/fn.read.html) (albeit with the added safety of taking a shared reference rather than a raw pointer, however the former coerces to the latter so it makes little ergonomic difference)?

Right, that's what I was looking for, thanks. I remembered std had something for that but didn't recall it was in ptr not in mem...
